### PR TITLE
Support for multiple signing certificates in downloaded metadata (#105) + Fix for ValidateInResponseTo results in "The Saml2Response didn't pass validation" in a farm setup #98

### PR DESCRIPTION
**Pull request by doormalena**
Exisiting code is altered to fix issue reported in #190.  
- Kentor.AuthServices/IdentityProvider.cs
- Kentor.AuthServices/SAML2P/Saml2Response.cs

Fix for ValidateInResponseTo results in "The Saml2Response didn't pass validation" in a farm setup #98
- Kentor.AuthServices/Internal/PendingAuthnRequests.cs

### CODE REVIEW

**Comment by albinsunnanbo** [comment]
Is the ReturnUrl support removed?

**Comment by doormalena** [comment]
Thanks for the comment. I will have a look at it on monday. It was not intentionally removed.


-------- Original message --------
From: Albin Sunnanbo
Date:14/02/2015 08:41 (GMT+01:00)
To: KentorIT/authservices
Cc: Alexander van Doormalen
Subject: Re: [authservices] Support for multiple signing certificates in downloaded metadata (#105) + Fix for ValidateInResponseTo results in "The Saml2Response didn't pass validation" in a farm setup #98 (#194)


Is the ReturnUrl support removed?

File: Kentor.AuthServices/Internal/PendingAuthnRequests.cs [hunk 1]
```diff
@@ -1,40 +1,94 @@
...
+        var cookie = HttpContext.Current.Request.Cookies[CookieName];
+
+        if (cookie != null)
+        {
+          var cookieId = Decrypt(cookie.Values[CookieIdAttribute]);
+          var cookieIdpId = Decrypt(cookie.Values[CookieIdpIdAttribute]);
+
+          if (id != null && cookieId == id.Value)
+          {
+            idp = new StoredRequestState(new EntityId(cookieIdpId), null); // TODO: returnUrl is left out here as it's not used. Possibly add it by also storing it into the cookie (or get it from config based in the id) or simply remove if from the stored state
```
Comment (albinsunnanbo) [inline Kentor.AuthServices/Internal/PendingAuthnRequests.cs]: Here ReturnUrl seems to be lost?

File: Kentor.AuthServices/Internal/PendingAuthnRequests.cs [hunk 2]
```diff
@@ -1,40 +1,94 @@
...
+      private const string CookieName = "AuthNRequest";
+      private const string CookieIdAttribute = "id";
+      private const string CookieIdpIdAttribute = "idpId";
+      private const string Purpose = "Initiator Verification";
+
+      internal static void Add(Saml2Id id, StoredRequestState idp)
+      {
+        var cookie = GetNewCookie();
+        cookie.Values[CookieIdAttribute] = Encrypt(id.Value);
+        cookie.Values[CookieIdpIdAttribute] = Encrypt(idp.Idp.Id);
```
Comment (albinsunnanbo) [inline Kentor.AuthServices/Internal/PendingAuthnRequests.cs]: Here ReturnUrl seems to be lost?
RelayData too.
Serialization of the whole StoredRequestState object might be a better solution?

Comment (doormalena) [inline Kentor.AuthServices/Internal/PendingAuthnRequests.cs]: Yes seems like a good proposal, also to support any additions/changes being done to the StoredRequestState. Thnx.

File: Kentor.AuthServices/Internal/PendingAuthnRequests.cs [hunk 3]
```diff
@@ -1,40 +1,94 @@
...
+        {
+          try
+          {
+            var protectedBytes = Convert.FromBase64String(value);
+            var unprotectedBytes = MachineKey.Unprotect(protectedBytes, Purpose);
+            returnValue = unprotectedBytes != null ? Encoding.UTF8.GetString(unprotectedBytes) : String.Empty;
+          }
+          catch (Exception)
+          {
+            // TODO: logging / exception. Probable causes: tempering with cookie, or machine keys are not equal on each farm member
```
Comment (albinsunnanbo) [inline Kentor.AuthServices/Internal/PendingAuthnRequests.cs]: TODO...

File: Kentor.AuthServices/IdentityProvider.cs [hunk 4]
```diff
@@ -76,7 +77,7 @@ private void Validate()
                 throw new ConfigurationErrorsException("Missing binding configuration on Idp " + EntityId.Id + ".");
             }
 
-            if (SigningKey == null)
+            if (SigningKeys == null)
```
Comment (albinsunnanbo) [inline Kentor.AuthServices/IdentityProvider.cs]: I suppose !SigningKeys.Any() should also throw configuration error?

Comment (doormalena) [inline Kentor.AuthServices/IdentityProvider.cs]: As it's now a list, null check indeed became invalid. Needs to be changed to SigningKeys.Count == 0
Will do this on Monday as well.

File: Kentor.AuthServices/SAML2P/Saml2Response.cs [hunk 5]
```diff
@@ -1,480 +1,487 @@
...
+                if (!allowedTransforms.Contains(transform.Algorithm))
+                {
+                    throw new Saml2ResponseFailedValidationException(
+                        "Transform \"" + transform.Algorithm + "\" found in Xml signature is not allowed in SAML.");
+                }
+            }
+
+            var validSignature = false;
+
+            for (var i = 0; !validSignature && i < idpKeys.Count; i++)
```
Comment (albinsunnanbo) [inline Kentor.AuthServices/SAML2P/Saml2Response.cs]: or

    var validSignature = idpKeys.Any(key => signedXml.CheckSignature(key));

or even

    var validSignature = idpKeys.Any(signedXml.CheckSignature);

makes it more compact and readable

File: Kentor.AuthServices/Internal/PendingAuthnRequests.cs (hunk 3 above)
Comment (doormalena) [inline Kentor.AuthServices/Internal/PendingAuthnRequests.cs]: What's best practice for this in the AuthServices Library? Throw exception?

**Comment by AndersAbel** [comment]
Thanks to both of you for a good discussion on the code in the Pull Request. Unfortunately, I have something that will cause some extra work to say: I want this PR to be split in two.

* One PR for the support for multiple keys (#105)
* One PR for the farm setup (#98 and #182)

Regarding the latter, I have been thinking quite a bit about how to approach the farm setup issue. It is a complex area, especially as the `StoredRequestState` provides protection against both replay attacks and XSRF attacks.

**Comment by albinsunnanbo** [comment]
Hmmm, can replay attacks be prevented by an encrypted cookie? If you replay the cookie too the server will happily decrypt it and find it valid.
The replay state needs to be stored on the server side.
I can see several strategies for sharing state in a web farm.

0) Current local dictionary for simple setup in non farm solutions
1) Dedicated state server
2) Redis or some other key value store
3) Sync state between servers using the windows service bus

We might need some plugin strategy here.

**Comment by doormalena** [comment]
If you want to prevent replay attacks, you also need to take into account the WIF created session. How is WIF protected against replay attacks? We can limit the replay window by adding a timestamp. 

If you don't want a cookie, we can also serialize, encrypt and base64 encode the StoredRequestState and send it as a RelayState parameter to the IDP. This however, doesn't prevent the replay, only prevents cookie re-usage.

A replay attack (after stealing the cookie) can only be prevent in this case by like we discussed earlier a server-side kept state (which can be a database, state server, shared cache mechanisme etc.). As discussed in #98, this makes the deployment a bit more complex.

Secondly, what about the AllowUnsolicitedAuthnResponse setting? If this is turned on, there is no verification at all.

**Comment by doormalena** [comment]
@albinsunnanbo 
I have processed all your feedback. Due note that the cookie mechanisme does have it's downsides which need to be considered (other than possible replay). A cookie can only store 4k of information. Adding the relaystate to it could potentionally break this construction. Adding the relaystate to the IDP requests would be a better solution or a state based approach (DB, state server etc...).

**Comment by AndersAbel** [comment]
Discussed this issue, and will close it as Alexander will create two separate pull requests for these features.

**Pull request closed without merging**